File: src/main/java/me/kuku/yuq/controller/BeforeController.java
package me.kuku.yuq.controller;

import com.IceCreamQAQ.Yu.annotation.Before;
import com.IceCreamQAQ.Yu.annotation.Catch;
import com.IceCreamQAQ.Yu.annotation.Global;
import com.IceCreamQAQ.Yu.cache.EhcacheHelp;
import com.IceCreamQAQ.Yu.entity.DoNone;
import com.alibaba.fastjson.JSONObject;
import com.icecreamqaq.yuq.FunKt;
import com.icecreamqaq.yuq.annotation.GroupController;
import com.icecreamqaq.yuq.annotation.PrivateController;
import com.icecreamqaq.yuq.controller.BotActionContext;
import com.icecreamqaq.yuq.message.Message;
import kotlinx.coroutines.TimeoutCancellationException;
import me.kuku.yuq.entity.GroupEntity;
import me.kuku.yuq.entity.GroupService;
import me.kuku.yuq.entity.QqEntity;
import me.kuku.yuq.entity.QqService;
import me.kuku.yuq.logic.ToolLogic;
import net.mamoe.mirai.contact.BotIsBeingMutedException;

import javax.inject.Inject;
import javax.inject.Named;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;

@GroupController
@PrivateController
public class BeforeController {

	@Inject
	private QqService qqService;
	@Inject
	private GroupService groupService;
	@Inject
	@Named("CommandCountOnTime")
	private EhcacheHelp<Integer> eh;
	@Inject
	private ToolLogic toolLogic;

	@Global
	@Before(weight = -1)
	public void nextParams(BotActionContext context, Long qq, Long group){
		QqEntity qqEntity = qqService.findByQq(qq);
		context.set("qqEntity", qqEntity);
		if (group != null){
			GroupEntity groupEntity = groupService.findByGroup(group);
			context.set("groupEntity", groupEntity);
		}
	}

	@Global
	@Before
	public void before(Message message, Long group, Long qq){
		if (group == null) return;
		List<String> list = message.toPath();
		if (list.size() == 0) return;
		String command = list.get(0);
		if (command.equals("指令限制") || command.equals("加指令限制")) return;
		GroupEntity groupEntity = groupService.findByGroup(group);
		if (groupEntity == null) return;
		Integer maxCount = groupEntity.getMaxCommandCountOnTime();
		if (maxCount == null) maxCount = -1;
		if (maxCount < 0) return;
		String key = "qq" + qq.toString() + command;
		Integer num = eh.get(key);
		if (num == null) num = 0;
		if (num >= maxCount) throw new DoNone();
		eh.set(key, ++num);
	}

	@Global
	@Before
	public void before(Message message, Long group){
		if (group == null) return;
		List<String> list = message.toPath();
		if (list.size() == 0) return;
		GroupEntity groupEntity = groupService.findByGroup(group);
		if (groupEntity == null) return;
		JSONObject jsonObject = groupEntity.getCommandLimitJson();
		String command = list.get(0);
		if (jsonObject.containsKey(command)){
			Integer maxCount = jsonObject.getInteger(command);
			if (maxCount < 0) return;
			String key = "group" + group.toString() + command;
			Integer num = eh.get(key);
			if (num == null) num = 0;
			if (num >= maxCount) throw new DoNone();
			eh.set(key, ++num);
		}
	}

	@Global
	@Catch(error = Exception.class)
	public void recording(Exception exception, long group, long qq){
		if (exception instanceof TimeoutCancellationException
				|| exception instanceof IOException || exception instanceof BotIsBeingMutedException){
			return;
		}
		StringWriter sw = new StringWriter();
		PrintWriter pw = new PrintWriter(sw);
		exception.printStackTrace(pw);
		try {
			String url = toolLogic.pasteUbuntu("exception", "java", sw.toString());
			FunKt.getYuq().getGroups().get(group).sendMessage(FunKt.getMif().at(qq).plus("程序出现异常了，异常如下：" + url));
		}catch (Exception ignore){
		}
	}

	@Global
	@Catch(error = IOException.class)
	public void interIO(IOException iOException, long qq, long group){
		FunKt.getYuq().getGroups().get(group).sendMessage(FunKt.getMif().at(qq).plus("出现io异常了，请重试！！"));
	}

	@Global
	@Catch(error = BotIsBeingMutedException.class)
	public void innerMuted(BotIsBeingMutedException e, long qq, long group){
		FunKt.getYuq().getGroups().get(group).get(qq).sendMessage(Message.Companion.toMessage("机器人被禁言了，不能发送消息啦！！"));
	}

}

File: src/main/kotlin/me/kuku/yuq/controller/BeforeController.kt
@file:Suppress("UNUSED_PARAMETER")

package me.kuku.yuq.controller

import com.IceCreamQAQ.Yu.annotation.After
import com.IceCreamQAQ.Yu.annotation.Before
import com.IceCreamQAQ.Yu.annotation.Catch
import com.IceCreamQAQ.Yu.annotation.Global
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.annotation.PrivateController
import com.icecreamqaq.yuq.controller.BotActionContext
import com.icecreamqaq.yuq.controller.ContextSession
import com.icecreamqaq.yuq.entity.Friend
import com.icecreamqaq.yuq.entity.Group
import com.icecreamqaq.yuq.entity.Member
import com.icecreamqaq.yuq.entity.MessageAt
import com.icecreamqaq.yuq.error.WaitNextMessageTimeoutException
import com.icecreamqaq.yuq.message.*
import com.icecreamqaq.yuq.mif
import kotlinx.coroutines.runBlocking
import me.kuku.utils.JobManager
import me.kuku.yuq.config.VerificationFailureException
import me.kuku.yuq.entity.*
import org.springframework.stereotype.Component

@PrivateController
@GroupController
@Component
class BeforeController (
    private val qqService: QqService,
    private val groupService: GroupService,
    private val messageService: MessageService,
    private val privateMessageService: PrivateMessageService,
    private val exceptionLogService: ExceptionLogService
){

    @Before(weight = -1)
    @Global
    fun before(session: ContextSession, qq: Long, group: Long?) {
        val qqEntity = qqService.findByQq(qq)
        session["qqEntity"] = qqEntity!!
        if (group != null) {
            val groupEntity = groupService.findByGroup(group)
            session["groupEntity"] = groupEntity!!
        }
    }

    @After
    @Global
    fun after(context: BotActionContext, qq: Long) {
        context.reMessage?.let { message ->
            val source = context.source
            if (source is Member || source is Group) {
                if (message.at == null && message.body.find { it is At } == null) {
                    message.at = MessageAt(qq, true)
                }
            }
        }
    }

    @Catch(error = WaitNextMessageTimeoutException::class)
    @Global
    fun waitError(exception: WaitNextMessageTimeoutException, qq: Long) {
        throw mif.at(qq).plus("等待您的消息超时，上下文已结束").toThrowable()
    }

    @Catch(error = VerificationFailureException::class)
    @Global
    fun ss(exception: VerificationFailureException, qq: Long) {
        throw mif.at(qq).plus(exception.message).toThrowable()
    }

    @Catch(error = Exception::class)
    @Global
    fun ss(exception: Exception, message: Message, context: BotActionContext, qq: Long, group: Long?) {
        if (exception is WaitNextMessageTimeoutException || exception is VerificationFailureException) return
        val exceptionStackTrace = exception.stackTraceToString()
        val url = runBlocking {
            exception.toUrl()
        }
        val source = context.source
        source.sendMessage(mif.at(qq).plus("程序出现异常了，异常信息为：$url"))
        JobManager.now {
            exception.save(url)
        }
        val messageId = message.source?.id ?: 0
        if (source is Friend || source is Member) {
            val messageEntity = privateMessageService.findByMessageIdAndQq(messageId, qq) ?: return
            val exceptionLogEntity = ExceptionLogEntity()
            exceptionLogEntity.privateMessageEntity = messageEntity
            exceptionLogEntity.stackTrace = exceptionStackTrace
            exceptionLogEntity.url = url
            exceptionLogService.save(exceptionLogEntity)
        } else if (source is Group) {
            val messageEntity =  messageService.findByMessageIdAndGroup(messageId, group!!) ?: return
            val exceptionLogEntity = ExceptionLogEntity()
            exceptionLogEntity.messageEntity = messageEntity
            exceptionLogEntity.stackTrace = exceptionStackTrace
            exceptionLogEntity.url = url
            exceptionLogService.save(exceptionLogEntity)
        }
    }

}
